From: Ondrej Zary <linux@zary.sk>
Date: Thu, 5 Oct 2023 22:55:59 +0200
Subject: ata: pata_parport: fit3: implement IDE command set registers
Patch-mainline: v6.6-rc6
Git-commit: 0c1e81d0b5ebd5813536dd5fcf5966ad043f37dc
References: git-fixes

fit3 protocol driver does not support accessing IDE control registers
(device control/altstatus). The DOS driver does not use these registers
either (as observed from DOSEMU trace). But the HW seems to be capable
of accessing these registers - I simply tried bit 3 and it works!

The control register is required to properly reset ATAPI devices or
they will be detected only once (after a power cycle).

Tested with EXP Computer CD-865 with MC-1285B EPP cable and
TransDisk 3000.

Signed-off-by: Ondrej Zary <linux@zary.sk>
Reviewed-by: Sergey Shtylyov <s.shtylyov@omp.ru>
Signed-off-by: Damien Le Moal <dlemoal@kernel.org>
Acked-by: Martin Wilck <mwilck@suse.com>
---
 drivers/ata/pata_parport/fit3.c |   15 +++------------
 1 file changed, 3 insertions(+), 12 deletions(-)

--- a/drivers/ata/pata_parport/fit3.c
+++ b/drivers/ata/pata_parport/fit3.c
@@ -9,12 +9,6 @@
 
 	The TD-2000 and certain older devices use a different protocol.
 	Try the fit2 protocol module with them.
-
-        NB:  The FIT adapters do not appear to support the control 
-	registers.  So, we map ALT_STATUS to STATUS and NO-OP writes 
-	to the device control register - this means that IDE reset 
-	will not work on these devices.
-
 */
 
 #include <linux/module.h>
@@ -37,8 +31,8 @@
 */
 
 static void fit3_write_regr(struct pi_adapter *pi, int cont, int regr, int val)
-
-{	if (cont == 1) return;
+{
+	regr += cont << 3;
 
 	switch (pi->mode) {
 
@@ -60,10 +54,7 @@ static int fit3_read_regr(struct pi_adap
 
 {	int  a, b;
 
-	if (cont) {
-	  if (regr != 6) return 0xff;
-	  regr = 7;
-	} 
+	regr += cont << 3;
 
 	switch (pi->mode) {
 
